Modify auth flow for transkribus to include refresh token

Parthiv-M opened the pull request:
Modify the authentication flow of the Transkribus engine to use refresh tokens in addition to access tokens

Bug: T330052

Commit:
Prompt user to also store refresh token in the app:transkribus command

Commit:
Update instructions in CONTIRBUTING.md for refresh tokens

samwilson commented on src/Command/TranskribusAuthCommand.php:
```diff
@@ -88,11 +88,18 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+                        $io->writeln('Your refresh token is: '.$refreshToken);
+                        $io->writeln('');
+                        $io->writeln('--- Please copy and add it to your .env.local'.
+                                                ' as APP_TRANSKRIBUS_REFRESH_TOKEN ---');
```
It might be good to output exactly what needs to be copied into `.env.local`, so we don't have to copy and paste twice.

Commit:
refactor CLI output for ease of copying

kolakachi committed:
Include refresh token in Transkribus engine

kolakachi committed:
fix linting errors

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -137,8 +146,52 @@ private function request(
+
+    private function refreshAccessToken(): void
+    {
+        $response = $this->httpClient->request(
```
This request is identical to the one done in `TranskribusAuthCommand` isn't it? I wonder if it'd be neater to inject a `TranskribusClient` into the command, and so only have this code in one place?

kolakachi replied:
New static methods have been added.

samwilson replied:
The method shouldn't be static. A TranskribusClient object should be injected in the constructor of the command, and then the method used non-statically. Does that make sense?

kolakachi replied:
Static methods to avoid injecting empty accessToken/refereshToken when running the command for the first time. Updates have been made. The methods are no longer static. The client no longer requires the accesstoken and refreshtoken when it's created.

samwilson commented on tests/OcrTestCase.php:
```diff
@@ -13,10 +13,14 @@ class OcrTestCase extends KernelTestCase
         self::bootKernel();
         $this->projectDir = self::$kernel->getProjectDir();
         $this->accessToken = "eyJhbGciOiJSUzI1NiIsInR5cC";
+        $this->refreshToken = "eyJhbGciOiJSUzI1NiIsInR5cC";
```
Sorry, I know I should look this up, but shouldn't test env vars like these be defined in `phunit.xml.dist`? If not, no worries.

Also, it might be good to make these different values, and comprehensible, so people don't think they're meant to be identical or useful. e.g. `eyThisIsATestAccessToken`.

samwilson replied:
Do they not work when added to `phpunit.xml.dist`?

kolakachi committed:
change OCRTestCase refreshToken value

kolakachi committed:
Refactor TranskribusClient

Added static methods for making accesstoken and refreshtoken requests

kolakachi committed:
fix linting errors

kolakachi committed:
fix linting error

kolakachi committed:
make OCRTestCase accessToken and refreshToken comprehensible

Parthiv-M commented on src/Engine/TranskribusClient.php:
```diff
@@ -137,8 +146,97 @@ private function request(
+        return $response;
+    }
+
+    /**
+     * @param string $userName
+     * @param string $password
+     * @param HttpClientInterface $client
+     */
+    public static function getAccessTokenResponse(
+        string $userName,
+        string $password,
+        HttpClientInterface $client
+    ): object {
+        $body = [
```
I believe this function is not being called anywhere. Can we remove it?

kolakachi replied:
The Intent is for it to be used in the TranskribusAuthCommand class to have the requests in one place

kolakachi committed:
Refactor TranskribusClient
Removed accessToken and refreshToken arguments from TranskribusClient constructor
Added env variables to phpunit.xml.dist

kolakachi committed:
Refactor logic for setting access token

Commit:
Update TranskribusAuthCommand to use functions from TranskribusClient and removed reduntant auth code

samwilson commented on tests/Controller/OcrControllerTest.php:
```diff
@@ -47,7 +47,9 @@ public function testGetLang(array $getParams, array $expectedLangs): void
-                    new TranskribusClient(new MockHttpClient(), $this->accessToken),
+                    new TranskribusClient(new MockHttpClient()),
+                    $_ENV['APP_TRANSKRIBUS_ACCESS_TOKEN'],
+                    $_ENV['APP_TRANSKRIBUS_REFRESH_TOKEN'],
```
These should use `getenv()`, to avoid use of globals.

samwilson commented on config/services.yaml:
```diff
@@ -44,10 +44,11 @@ services:
             $accessToken: '%env(APP_TRANSKRIBUS_ACCESS_TOKEN)%'
+            $refreshToken: '%env(APP_TRANSKRIBUS_REFRESH_TOKEN)%'
+
+    App\Engine\TranskribusClient:
```
Is this empty config needed?

samwilson commented on src/Command/TranskribusAuthCommand.php:
```diff
@@ -15,19 +16,37 @@ class TranskribusAuthCommand extends Command
     /** @var HttpClientInterface The interface to make http requests */
     private $client;
```
This is no longer used.

samwilson commented on src/Command/TranskribusAuthCommand.php:
```diff
@@ -15,19 +16,37 @@ class TranskribusAuthCommand extends Command
+     * @param TranskribusClient $transkribusClient
+     */
+    public function __construct(
+        HttpClientInterface $client,
```
This service can be removed now, I think?

samwilson commented on tests/OcrTestCase.php:
```diff
@@ -13,10 +13,12 @@ class OcrTestCase extends KernelTestCase
     protected $accessToken;
 
+    /** @var string */
+    protected $refreshToken;
```
This is not used (and $accessToken can also be removed).

samwilson commented on src/Engine/TranskribusClient.php:
```diff
@@ -137,8 +149,89 @@ private function request(
+                if (!empty($content)) {
+                    return $content;
+                }
+                $this->throwException(0);
```
I know it's not being introduced in this patch, but these magic numbers are not great I think. Better to set the codes to be class constants and then use them here, e.g. `$this->throwException(self::ERROR_NO_CONTENT);` (can be done in a separate patch of course).

samwilson reviewed (commented):
Looks great, works in my tests. Just a few errant variables that are no longer required, but otherwise good to go I think.

Commit:
Removed unused variables from TranskribusAuthCommand

Commit:
Remove unused import from TranskribusAuthCommand

kolakachi committed:
Remove global variables from OcrControllerTest

kolakachi committed:
Update TranskribusClient class and services.yaml

Replaced magic number with a constant in TranskribusClient
and removed TranskribusClient services from services config

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -22,13 +22,16 @@ class TranskribusEngine extends EngineBase
         parent::__construct($intuition, $projectDir, $httpClient);
 
         $this->transkribusClient = $transkribusClient;
+        $this->transkribusClient->setTokens($accessToken, $refreshToken);
```
The tokens are only injected here in order to be passed through to the client — would it be better to inject them into the client instead?

kolakachi committed:
Refactor TranskribusClient and TranskribusEngine

Moved token injection from TranskribusEngine to TranskribusClient
Updated OcrControllerTest and services.yaml
